### PR TITLE
Implement NJT embedding backward

### LINKED ISSUE #138352: NJT Embedding backward
### 🚀 The feature, motivation and pitch

Implement `embedding_dense_backward` for nested jagged tensors. We can land this in `torch/nested/_internal/ops.py` using the `register_jagged_func` API.

### Alternatives

_No response_

### Additional context

Here's how I have this partially implemented in my project:

```python
# jagged.py
import torch
from torch.nested._internal.ops import check_ragged_dim_same, normalize_function, register_jagged_func

@register_jagged_func(
    torch.ops.aten.embedding_dense_backward.default,
    "self: jt, grad_output: jt, num_weights: any, padding_idx: any, scale_grad_by_freq: any",
)
def embedding_dense_backward(func, *args, **kwargs):
    _, new_kwargs = normalize_function(  # type: ignore[misc]
        func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
    )

    indices: NestedTensor = new_kwargs.pop("indices")
    num_weights: int = new_kwargs.pop("num_weights")
    grad_output: NestedTensor = new_kwargs.pop("grad_output")
    if new_kwargs["padding_idx"] != -1 or new_kwargs["scale_grad_by_freq"]:
        raise NotImplementedError("Haven't done this yet")

    check_ragged_dim_same(func, indices, "self", grad_output, "grad_output")
    out = torch.zeros(num_weights, grad_output.size(-1))
    src = grad_output._values
    indices = indices._values.long().unsqueeze(1).expand(-1, src.size(1))
    out.scatter_add_(dim=0, index=indices, src=src)
    return out
```

And here's the unit test I wrote for this:
```python
# test_jagged.py
import torch
from torch.nested._internal.nested_tensor import jagged_from_list

def test_Embedding_parity() -> None:
    import .jagged  # noqa: F401

    expected = torch.tensor([3, 1, 0, 0, 0], dtype=torch.get_default_dtype()).unsqueeze(1).repeat(1, 6)

    emb = torch.nn.Embedding(5, 6)
    ix = torch.tensor([0, 0, 0, 1], dtype=torch.int32)
    emb(ix).sum().backward()
    g1 = emb.weight.grad
    emb.weight.grad = None

    ix, _ = jagged_from_list([ix], offsets=None, dtype=torch.int32)
    emb(ix).values().sum().backward()
    g2 = emb.weight.grad

    torch.testing.assert_close(g1, expected)
    torch.testing.assert_close(g2, expected)
    torch.testing.assert_close(g1, g2)
```

cc @cpuhrsch @jbschlosser @bhosmer @drisspg @soulitzer @davidberard98 @YuqingJ

### PR DESCRIPTION
**Pull request by schmidt-jake**
Fixes #138352

### CODE REVIEW

File: test/test_nestedtensor.py [hunk 1]
```diff
@@ -1053,6 +1053,7 @@ def test_embedding(self, device, layout):
         )
         emb = torch.nn.Embedding(100, 8, device=device)
         y = emb(x)
+        y.sum().backward()
```
Comment (schmidt-jake) [inline test/test_nestedtensor.py:1056]: Is this good enough? Should I add a check for the gradients?

File: torch/nested/_internal/ops.py [hunk 2]
```diff
@@ -1678,6 +1678,31 @@ def embedding_default(func, *args, **kwargs):
...
+def embedding_dense_backward(func, *args, **kwargs):
+    _, new_kwargs = normalize_function(  # type: ignore[misc]
+        func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+    )
+
+    indices: NestedTensor = new_kwargs.pop("indices")
+    num_weights: int = new_kwargs.pop("num_weights")
+    grad_output: NestedTensor = new_kwargs.pop("grad_output")
+    if new_kwargs["scale_grad_by_freq"]:
+        raise NotImplementedError
```
Comment (schmidt-jake) [inline torch/nested/_internal/ops.py:1694]: Happy to try adding support for when `scale_grad_by_freq` is `True`...

File: torch/nested/_internal/ops.py [hunk 3]
```diff
@@ -1678,6 +1678,31 @@ def embedding_default(func, *args, **kwargs):
...
+
+    check_ragged_dim_same(func, indices, "self", grad_output, "grad_output")
+    out = torch.zeros(num_weights, grad_output.size(-1))
+    src = grad_output._values
+    indices = indices._values
+    indices = indices[indices != new_kwargs["padding_idx"]]
+    indices = indices.long().unsqueeze(1).expand(-1, src.size(1))
+    out.scatter_add_(dim=0, index=indices, src=src)
+    return out
```
Comment (schmidt-jake) [inline torch/nested/_internal/ops.py:1703]: Not sure this is the best way to compute the gradient. Open to suggestions!

File: torch/nested/_internal/ops.py [hunk 4]
```diff
@@ -1678,6 +1678,31 @@ def embedding_default(func, *args, **kwargs):
...
+    if new_kwargs["scale_grad_by_freq"]:
+        raise NotImplementedError
+
+    check_ragged_dim_same(func, indices, "self", grad_output, "grad_output")
+    out = torch.zeros(num_weights, grad_output.size(-1))
+    src = grad_output._values
+    indices = indices._values
+    indices = indices[indices != new_kwargs["padding_idx"]]
+    indices = indices.long().unsqueeze(1).expand(-1, src.size(1))
+    out.scatter_add_(dim=0, index=indices, src=src)
```
Comment (jbschlosser) [inline torch/nested/_internal/ops.py:1703]: Rather than reimplement the backward behavior manually, I think it's possible to simply redispatch to the dense implementation (i.e. call `func` with unwrapped buffers). I think this should do the trick:
```python
@register_jagged_func(
    torch.ops.aten.embedding_dense_backward.default,
    "grad_output: jt, indices: jt, num_weights: any, padding_idx: any, scale_grad_by_freq: any",
)
def embedding_dense_backward_default(func, *args, **kwargs):
    _, new_kwargs = normalize_function(  # type: ignore[misc]
        func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
    )

    indices = new_kwargs.pop("indices")
    grad_output = new_kwargs.pop("grad_output")
    return func(grad_output._values, indices._values, **new_kwargs)
```

File: test/test_nestedtensor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jbschlosser) [reply test/test_nestedtensor.py:1056]: Note that I recently landed some convenient testing infrastructure for NJT + OpInfo entries. You can take advantage of this and get forward / backward / eager / compile test coverage automagically via:
```diff
diff --git a/torch/testing/_internal/opinfo/definitions/nested.py b/torch/testing/_internal/opinfo/definitions/nested.py
index f13caa8c49e..58b8e61c971 100644
--- a/torch/testing/_internal/opinfo/definitions/nested.py
+++ b/torch/testing/_internal/opinfo/definitions/nested.py
@@ -417,6 +417,38 @@ def sample_inputs_masked_select(
         )


+def sample_inputs_nn_functional_embedding(
+    op_info, device, dtype, requires_grad, **kwargs
+):
+    indices = torch.nested.nested_tensor(
+        [
+            torch.tensor([0, 2, 1, 3]),
+            torch.tensor([4, 2, 1]),
+            torch.tensor([6, 7, 5, 2, 4]),
+        ],
+        layout=torch.jagged,
+        dtype=torch.int64,
+        device=device,
+    )
+
+    NUM_EMBEDDINGS = 20
+    EMBEDDING_DIM = 32
+    weight = torch.randn(NUM_EMBEDDINGS, EMBEDDING_DIM, device=device, dtype=dtype)
+
+    # NB: the OpInfo entry for embedding_bag expects weight first so the gradients
+    # can be checked
+    yield SampleInput(
+        weight.clone().detach().requires_grad_(),
+        args=(indices,),
+    )
+
+    yield SampleInput(
+        weight.clone().detach().requires_grad_(),
+        args=(indices,),
+        kwargs={"padding_idx": 1},
+    )
+
+
 def sample_inputs_nn_functional_embedding_bag(
     op_info, device, dtype, requires_grad, **kwargs
 ):
@@ -549,6 +581,7 @@ njt_sample_inputs = {
     "bmm": sample_inputs_bmm,
     "clone": sample_inputs_clone,
     **{f"mvlgamma.mvlgamma_p_{p}": sample_inputs_mvl_gamma(p=1) for p in (1, 3, 5)},
+    "nn.functional.embedding": sample_inputs_nn_functional_embedding,
     "nn.functional.embedding_bag": sample_inputs_nn_functional_embedding_bag,
     "nn.functional.linear": sample_inputs_nn_functional_linear,
     "nn.functional.rms_norm": sample_inputs_nn_functional_rms_norm,
```

**Review by jbschlosser — COMMENTED** [review]
Hey thanks so much for the PR! Some testing and implementation comments below :)

File: test/test_nestedtensor.py (hunk 1 above)
Comment (jbschlosser) [reply test/test_nestedtensor.py:1056]: Hey @schmidt-jake - the testing comment I left involving NJT + OpInfo entries should give you gradient checking automatically! lmk if you have questions

File: torch/nested/_internal/ops.py (hunk 2 above)
Comment (jbschlosser) [reply torch/nested/_internal/ops.py]: I think we'll get this for free by redispatching onto the underlying `func` :) (see my comment below)

**Comment by schmidt-jake** [comment]
@jbschlosser My mind is blown! Thanks for the help.

**Review by jbschlosser — APPROVED** [review]
Looks great! Thanks again :)
